### ISSUE TITLE
Bug (compactor):  Investigate compactor oom bug and improve memory allocation

### ISSUE THREAD

**Issue by Li0k** [risingwavelabs/risingwave#15022]
### Background
The phenomenon of compactor oom occurs during the compactor-test.
<img width="937" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/5108fe2b-1402-4c4c-9161-0cbef9db06b4">

We found that when the test is first started, compactor's memory spikes and experiences oom, and stabilizes after a while.

### Investigate
I turned on the memory profiler and reran the test, and found a few things.
<img width="1511" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/b9187332-2ecf-4343-aeb1-5873e7a38258">

After the introduction of check_compact_result, it was taking up some of the memory.  I found out that the memory tracker is not maintained during the execution of check_compact_result, which could lead to a memory overflow.

<img width="1012" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/353a3341-30bc-48f4-ba70-8f721e356979">

Unfortunately, this only delayed the appearance of oom and did not solve the problem. so continued to analyze the metrics and the profile.There was some interesting information in the metrics that helped me investigate the cause of the problem: Why is the oom only appearing at startup? A simple way to think about it is to look for fluctuations in the same time.

1. multi version key
<img width="952" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/b57f6d70-c84a-4c34-ac59-32646c9a8c39">

2. k-v pair size
<img width="966" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/f1091609-9c5c-44e2-bc82-c2276d3b2835">

3. epoch
<img width="945" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/c1c9266b-d357-45d1-8524-68ef6ebb3260">

According to the above metrics, we can find that in the startup phase of the test (10min), min_epoch is pinned, which leads to a large number of multi-version keys in the system, and the size of k-v pairs in the test is larger than that of a normal nexmark test.

Analyzing further, based on the profile results, we can locate the problem in the encode and decode phases of the block.
1. 
![image](https://github.com/risingwavelabs/risingwave/assets/15023748/fead2e63-fea1-42ec-8707-0c57b3348748)

2. 
![image](https://github.com/risingwavelabs/risingwave/assets/15023748/3034d6bf-03a9-4a17-a96f-3d8aacf62357)

3. 
![image](https://github.com/risingwavelabs/risingwave/assets/15023748/a2c1981f-e3a1-41fd-b161-b4df54efe986)

From the profile analysis, BlockBuilder::add consumes a lot of memory, based on the metrics above, it may be due to the effect of multiple key versions. Reading the code further, after https://github.com/risingwavelabs/risingwave/commit/8136111a7dac6034759952a40a3b457833152d57, the builder tends to place the same user_key in the block. 

<img width="995" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/83c9b8e8-8e57-47a2-a286-8bb023b062dd">

TLDR: the limit of is_new_user_key could lead to huge blocks under the current test, which are not covered by the memory estimation algorithm, and trigger more memory allocations.

**Comment by Li0k** [comment]
### Improvement
Based on the information in the Memory profile, we found some points that can be optimized.

1. BlockBuilder::Add

  ![image](https://github.com/risingwavelabs/risingwave/assets/15023748/855d09f1-88b2-43ba-a32f-b4abc48e316b)
  
  BlockBuilder preallocates the buf at startup to minimize secondary memory allocation. 
  
  <img width="689" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/d9f6bb98-ad11-4fa5-b517-89014cbcc5fa">
  
  There are two problems:
  - the extra 256 allocated may not be enough after the introduction of the type index, and a large number of keys will trigger an allocation of
  
  - During compress, the buf variable is changed, invalidating the expected pre-allocation.
  <img width="910" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/11aca127-6cfd-48d6-b22d-59feefbb0fc2">  


2. BlockBuilder::compress

  ![image](https://github.com/risingwavelabs/risingwave/assets/15023748/36c392c9-c653-41ac-b817-8a0025d3301a)
  
  compress function will generate a new BytesMut write each time for encoding
  
  <img width="969" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/a4b7109f-8047-4b6f-8364-be78a6c62498">

3. Block::decode_with_copy

![image](https://github.com/risingwavelabs/risingwave/assets/15023748/bb37c42e-8ba1-4bd2-9b7a-b1478788a854)

Crete Bytes from Vec lead to memory reallocation.

```rust
    /// Converts the vector into [`Box<[T]>`][owned slice].
    ///
    /// If the vector has excess capacity, its items will be moved into a
    /// newly-allocated buffer with exactly the right capacity.
    #[cfg(not(no_global_oom_handling))]
    #[stable(feature = "rust1", since = "1.0.0")]
    pub fn into_boxed_slice(mut self) -> Box<[T], A> {
        unsafe {
            self.shrink_to_fit();
            let me = ManuallyDrop::new(self);
            let buf = ptr::read(&me.buf);
            let len = me.len();
            buf.into_box(len).assume_init()
        }
    }
```

<img width="1276" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/9d1025e9-50a5-4a10-a37e-cf35239bba84">

#### Change
1. reserve more space for type index
2. change the condition for determining block full to avoid block size exceeding capacity as much as possible.
3. Introduce a dedicated compress buf
4. refactor the usage of Bytes

#### Result

1. 
<img width="1511" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/d01d46bc-768d-4e33-841f-d3ae8ed4371a">

2. 
<img width="1511" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/a7f26eb1-a3c2-41ae-9be7-3983adf1cde5">

3. 
<img width="1511" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/15023748/b836812b-3127-4aa8-9c0d-2fa6ef361aae">

**Comment by hzxa21** [comment]
> From the profile analysis, BlockBuilder::add consumes a lot of memory, based on the metrics above, it may be due to the effect of multiple key versions. Reading the code further, after https://github.com/risingwavelabs/risingwave/commit/8136111a7dac6034759952a40a3b457833152d57, the builder tends to place the same user_key in the block.

What is the benefit of putting all versions of a user key into a single block? I guess it is a bug?
